### ISSUE TITLE
Add example for non-trivial and stateful view/filter (e.g. XML entity decoder)

### ISSUE THREAD

**Issue by RokerHRO** [ericniebler/range-v3#1098]
I'd like to see how I can write a non-trivial view that has internal state that has to be kept between the iteration over the underlaying container and that produces 0…n octet output for a given input octet, depending on the input char and its internal state.

An example would be a base64 encoder/decoder or a decoder for XML entities (`&amp;` , `&lt`; and such).

Or is that doable only with combination of existing filters and views? If yes, an example should show that. :-)

**Comment by tower120** [comment]
> If yes, an example should show that. :-)

I think you ask too much :) 

View can have internal state - just keep that state inside view, not inside iterator (iterator would store pointer to view). Use `ranges::view_adaptor` - that would be the shortest way.

> Or is that doable only with combination of existing filters and views?

If you don't need to skip any elements - `view::transform` can work with statefull functors.

**Comment by RokerHRO** [comment]
Thank you for you quick reply.

But I still have no clue how to do a "base64-encoding" view: it has to "eat" 3 octets on its input and produce 4 octets to its output. Emit an additional "line break" after 72 output bytes (so the "line length" needs to be kept in its internal state). And needs some kind of "flush" when the input ends early (with only 1 or 2 consumed input octets).

Similar problems for a base64-decoding view: it has to read 4 "non-whitespace octets" (that means: skip any whitespace without counting them) and outputs 3 octets. And it has to emit erros somehow if it reads characters that cannot occur in valid base64-encoded input strings. :-/

And how I should keep the "state" in the view? Is it guaranteed that _every_ view is traversed _exactly_ _once_ and always traversed completely from `begin()` to `end()`, without any skipping or rewinding?

You're right, I ask a lot. I hope it is not too much, yet.

**Comment by tower120** [comment]
_DISCLAIMER: If you need it to just work, and you have access to compiler with co-routines support - just implement it as generator._

To do that with range-v3 views you probably need true generator to work (from coroutines).
Currently this probably would be the closest shot:
```cpp
using namespace ranges;

// this - could be generator with co_yield's
auto process_octet(auto& octet){
  // you probably can't just return container ... If can - thread_local is superfluous
  // with tread_local you can't have nested loops.
  thread_local boost::fixed_vector<T, 4> out_octets;
  out_octets.clear();

   if (??){
       out_octets = {1};
   }  else {
       out_octets = {2, 3};
   }

  return out_octets | view::all;  
}

auto vi =
   src 
   | view::for_each([](auto& octet){
        return yield_from(process_octet(octet));
    })
```

With your custom view:
 - To skip - just increase src range iterator. 
 - When generate extra element - do not increase src range iterator on `next()`.

This can be not an easy task.

> And how I should keep the "state" in the view?

You need `view_adaptor`, see in docs. In View store all your data, and in iterator have pointer to the view (you may look at view::transform implementation - that's how it work), thus you'll keep iterator small sized [but you can't have nested loops with this].

>  Is it guaranteed that every view is traversed exactly once and always traversed completely from begin() to end(), without any skipping or rewinding?

No. 

> I think, base64 is is well-known for most developers ...

It is not.

> You're right, I ask a lot. I hope it is not too much, yet.

You basically just ask to write you an XML parser.

**Comment by ericniebler** [comment]
> If you don't need to skip any elements - `view::transform` can work with statefull functors.

Sorry, no. Both filter and transform need `RegularInvocable`: when called with the same arguments, they must produce the same result. Otherwise, multiple traversals of the range will see different values, which breaks the semantics.

**Comment by tower120** [comment]
>  when called with the same arguments, they must produce the same result. 

Having internal state does not necessary means that functor will behave differently each run. Internal state can be some pre-computed cache map. Which probably is not what you'll have in parser...

**Comment by RokerHRO** [comment]
@tower120 : No, I don't ask you to write an XML parser. I just want to see an _example_  for a range/view that has internal state that has to be kept between several iteration steps.  

The suggested transformation of `&lt;`, `&gt;` and `&amp;` into `<`, `>` and `&` was just a simple idea (and is far from a real XML parser).  The base64 encoder/decoder was another idea for such example.

But if you prefer a different example, it is okay for me, too.

**Comment by tower120** [comment]
@RokerHRO I would like to see that example too, but I fear that currently, there is no such tools in range-v3 library that work out of box... You can make them probably...

Let me rephrase/simplify - you want something alike C# Enumerable / Rust Iterator? Something with this interface:

```cpp
struct MyParser : Enumerable<MyParser>{
    void rewind(){
        // start from begin
        // reset state here
    }
    std::optional<T> next(){
       // T - is value, not iterator
       // return empty - if this is the end
    }
}
```

`Enumerable` is Forward only, can be traversed **only** from begin to end, no "resume", only "current" iterator are valid. no nested traverses. _Optionally - can be traversed form begin again._ Something like InputIterator range.

?

**Comment by sv1990** [comment]
I found this an interesting problem and tried a solution. This is what I came up with. I'm sure it could be optimized but it seems to work. It is a range adaptor that can transform `&lt;` into `<` and so on...

    #include <iostream>
    #include <optional>
    #include <string>
    #include <unordered_map>

    #include <range/v3/all.hpp>
    #include <range/v3/iterator/stream_iterators.hpp>

    template <class Rng>
    class xml_decode_view : public ranges::view_adaptor<xml_decode_view<Rng>, Rng> {
      friend ranges::range_access;
      class adaptor : public ranges::adaptor_base {
        const xml_decode_view* _rng;
        ranges::range_value_type_t<Rng> _current;
        ranges::range_difference_type_t<Rng> _last_parsed_size;

        std::optional<std::pair<std::string, char>>
        parse_entity(ranges::iterator_t<Rng> it) const {
          static const std::unordered_map<std::string, char> entities{
              {"apos", '\''},
              {"amp", '&'},
              {"quot", '\"'},
              {"lt", '<'},
              {"gt", '>'}};
          std::string token =
              ranges::make_subrange(ranges::next(it), ranges::end(_rng->base())) |
              ranges::view::take(5) |
              ranges::view::take_while([](auto c) { return c != ';'; });
          if (auto entity = entities.find(token); entity != ranges::end(entities)) {
            return *entity;
          }
          else {
            return std::nullopt;
          }
        }

      public:
        adaptor() = default;
        adaptor(const xml_decode_view& rng) : _rng(&rng), _last_parsed_size(0) {
          auto it = ranges::begin(_rng->base());
          next(it);
        }
        char read(ranges::iterator_t<Rng> /*it*/) const { return _current; }

        // Advance the iterator and parse the next value
        void next(ranges::iterator_t<Rng>& it) {
          ranges::advance(it,                //
                          _last_parsed_size, //
                          ranges::end(_rng->base()));
          if (*it != '&') {
            _current            = *it;
            _last_parsed_size = 1;
          }
          else {
            if (auto entity = parse_entity(it); entity.has_value()) {
              _current          = entity->second;
              _last_parsed_size = ranges::distance(entity->first) + 2;
            }
            else {
              _current          = *it;
              _last_parsed_size = 1;
            }
          }
        }
      };
      adaptor begin_adaptor() const { return {*this}; }

    public:
      xml_decode_view() = default;
      xml_decode_view(Rng&& rng)
          : xml_decode_view::view_adaptor{std::forward<Rng>(rng)} {}
    };

    template <class Rng>
    xml_decode_view<Rng> xml_decode(Rng&& rng) {
       return {std::forward<Rng>(rng)};
    }

    int main() {
      std::string s = "&lt;&quot;foo &amp; bar&quot;&gt;"; // <"foo & bar">
      ranges::copy(xml_decode(s), ranges::ostream_iterator<>{std::cout});
      std::cout << '\n';
    }

**Comment by RokerHRO** [comment]
@sv1990: Boooy, that's a lot of code. :-o
I'll try to benchmark it whether that is worth, compared to a naïve `std::string xml_decode(const std::string)` function that fit in much less lines of code. :-(

**Comment by sv1990** [comment]
@RokerHRO I don't know if that is the best implementation for this...
Here is the equivalent view using coroutines:

    #include <iostream>
    #include <range/v3/all.hpp>
    #include <range/v3/experimental/utility/generator.hpp>
    #include <range/v3/iterator/stream_iterators.hpp>
    #include <unordered_map>
    
    template <typename Rng>
    ranges::experimental::generator<char> xml_decode(Rng&& rng) {
      static const std::unordered_map<std::string, char> entities{
          {"apos", '\''}, {"amp", '&'}, {"quot", '\"'}, {"lt", '<'}, {"gt", '>'}};
      for (auto it = ranges::begin(rng); it != ranges::end(rng); ++it) {
        if (*it != '&') {
          co_yield *it;
        }
        else {
          std::string token =
              ranges::make_subrange(ranges::next(it), ranges::end(rng)) |
              ranges::view::take(5) |
              ranges::view::take_while([](auto c) { return c != ';'; });
          if (auto entity = entities.find(token); entity != end(entities)) {
            ranges::advance(it, ranges::size(entity->first) + 1, ranges::end(rng));
            co_yield entity->second;
          }
          else {
            co_yield *it;
          }
        }
      }
    }
    
    auto xml_decode() {
      return ranges::make_pipeable([](auto&& rng) {
        using Rng = decltype(rng);
        return xml_decode(std::forward<Rng>(rng));
      });
    }
    
    int main() {
      std::string s = "&lt;&quot;foo &amp; bar&quot;&gt;";
      ranges::copy(s | ranges::view::filter([](auto c) { return c != ' '; }) |
                       xml_decode(),
                   ranges::ostream_iterator<>{std::cout});
      std::cout << '\n';
    }

**Comment by tower120** [comment]
>  compared to a naïve std::string xml_decode(const std::string)

@RokerHRO what IS this magic function?

> Here is the equivalent view using coroutines

@sv1990 did you work with coroutines? Does something like `xml_decode()` actually optimizes to inlined version? Or each `xml_decode()` call is a function call / jmp (plus small overhead on stack restore)? I heard Gor's talk about inlining this things. But I actually can hardly believe that non-trivial coroutine function like this can be inlined...

**Comment by sv1990** [comment]
@tower120 I haven't done anything real with coroutines yet. Only such small experiments. So I can't say anything meaningful about the optimization of this...

**Comment by RokerHRO** [comment]
@tower120 :  `std::string xml_decode(const std::string&)` is just the signature of a naïve (and inefficient?) entity-decoding function that iterates over the input string and creates a copy with the decoded output.

I'd write such a function like this:

```c++
#include <string>
#include <cstdint>
#include <iostream>

constexpr inline
uint64_t operator"" _case (const char* s, size_t len)
{
        return len ? operator""_case(s, len-1)*257u + uint8_t(s[len-1]) + 1u : 0u;
}

std::string xml_decode(const std::string& input)
{
    std::string ret; ret.reserve(size_t(input.size() * 1.11));
    std::uint64_t state = 0;
    for(char c : input)
    {
        if(state == 0 && c!='&')
        {
            ret += c;
            continue;
        }
        state = state*257 + uint8_t(c) + 1;
        switch(state)
        {
            case "&amp;"_case  : ret += '&'; state=0; break;
            case "&lt;"_case   : ret += '<'; state=0; break;
            case "&gt;"_case   : ret += '>'; state=0; break;
            case "&quot;"_case : ret += '"'; state=0; break;
#ifdef THROW_ON_UNKNOWN_ENTITY
            case "&"_case :
            case "&a"_case :
            case "&am"_case :
            case "&amp"_case :

            case "&l"_case:
            case "&lt"_case:

            case "&g"_case:
            case "&gt"_case:

            case "&q"_case:
            case "&qu"_case:
            case "&quo"_case:
            case "&quot"_case:
                break;
            
            default:
                throw std::runtime_error("Unknown entity in input");
#endif
        }
    }
    return ret;
}
```

So if the decoder can ignore illegal input strings it is quite short and simple. If it has to deal with illegal inputs (with unknown entities) the number of intermediate states would explode in this naïve approach, I know. ;-)

**Comment by aldanor** [comment]
After reading the docs, the examples, and this entire thread, it's still unclear to me what's the proper way to write a `view_adaptor` that is stateful. Could this be please added to the docs, as an idiomatic way of doing that? (Especially if the state is complex, e.g. can't be copied or moved, and/or if the values in the resulting view may contain references to this state)
